### ISSUE TITLE
Adopt ``Quantity`` type annotations that preserve units

### ISSUE THREAD

**Issue by namurphy** [PlasmaPy/PlasmaPy#1413]
For many of our functions, we use annotations for units:

```Python
def func(x: u.m, t: u.s) -> u.m / u.s:
    return (x / t).to(u.m / u.s)
```

These are not *type hint* annotations, since the expected value is a `Quantity` in most situations.  Instead the annotations are used by decorators like our validators and `astropy.units.quantity_input`.  Unfortunately, this means that we can't use static type checkers like [mypy](http://mypy-lang.org/) to check PlasmaPy code. 😿  Using mypy has the potential to help us find bugs, as well as errors in existing type hint annotations (#268).

Astropy now allows for [Quantity type annotations that preserve units](https://docs.astropy.org/en/stable/units/type_hints.html#preserving-units).  Both `typing.Union` and `typing.Optional` are supported.

```Python
def func(x: Quantity[u.kpc]) -> Quantity[u.m]:
    return x << u.m
```

We should migrate towards using `Quantity` annotations that preserve units in PlasmaPy.  However, before we do this, we would need to change our validators to make them work with this style of annotation.

**Comment by namurphy** [comment]
I closed the wrong issue here with #2346.  I meant to close #1414.

**Comment by namurphy** [comment]
Closed by #2421!
